cooperator/tests/test_cooperator.py

```diff
@@ -64,104 +74,94 @@ def _pay_invoice(self, invoice, payment_date=None):
-            share_line.share_number = 3
-
-        # test ir model access for cooperator coop user
-        self.as_cooperator_user()
-        read_request = self.browse_ref("cooperator.subscription_request_1_demo")
-        read_request.name = "test write request"
-        create_request = self.env["subscription.request"].create(request_values)
+            share_line_as_user.share_number = 3
+
+    @users("user-cooperator")
+    def test_cooperator_access_rules(self):
+        cooperator_user = self.ref("cooperator.res_users_user_cooperator_demo")
+        # cf comment in test_user_access_rules
+        resquest_as_cooperator = self.subscription_request_1.sudo(cooperator_user)
+        resquest_as_cooperator.name = "test write request"
+        create_values = self._get_dummy_subscription_requests_vals()
+        create_request = self.env["subscription.request"].create(create_values)
         with self.assertRaises(AccessError):
             create_request.unlink()
 
-        share_line = self.browse_ref("cooperator.share_line_1_demo")
-        share_line.share_number = 3
+        share_line_as_cooperator_user = self.share_line.sudo(cooperator_user)
+        share_line_as_cooperator_user.share_number = 3
         with self.assertRaises(AccessError):
-            share_line.unlink()
+            share_line_as_cooperator_user.unlink()
 
-        share_type = self.browse_ref("cooperator.product_template_share_type_1_demo")
-        share_type.list_price = 30
+        share_type_as_cooperator_user = self.share_x.sudo(cooperator_user)
+        share_type_as_cooperator_user.list_price = 30
         with self.assertRaises(AccessError):
             self.env["product.template"].create(
                 {
                     "name": "Part C - Client",
                     "short_name": "Part C",
                     "is_share": True,
-                    "default_share_product": True,
-                    "force_min_qty": True,
-                    "minimum_quantity": 2,
-                    "by_individual": True,
-                    "by_company": True,
                     "list_price": 50,
-                    "display_on_website": True,
                 }
             )
         with self.assertRaises(AccessError):
-            share_type.unlink()
-
-        # test ir model access for cooperator manager
-        self.as_cooperator_manager()
-        read_request = self.browse_ref("cooperator.subscription_request_1_demo")
-        read_request.name = "test write request"
-        create_request = self.env["subscription.request"].create(request_values)
+            share_type_as_cooperator_user.unlink()
+
+    @users("manager-cooperator")
+    def test_cooperator_manager_access_rules(self):
+        cooperator_manager = self.ref("cooperator.res_users_manager_cooperator_demo")
+        # cf comment in test_user_access_rules
+        request_as_cooperator_manager = self.subscription_request_1.sudo(
+            cooperator_manager
+        )
```

Comment:
What's the point of the `@users` decorator when you need to do all the cruft anyways?

Reply:
cf comment in test_user_access_rules :

```
# class object was loaded with root user and its rights
# so we need to reload it with demo user rights
```